docs/source/extending/overloading-guide.rst

```diff
@@ -0,0 +1,278 @@
+
+As you can see, the typing checking code has been increased significantly to
+match the new requirements. Also, multiple implementations---one for integers
+and one for floating-point---are provided. We check inside the typing scope
+which implementation should be used and also raise any custom error messages
+required. Importantly, the check for ``nan`` values is only present in the
+floating point implementation as this additional check creates a runtime
+overhead. This can easily be observed during profiling:
+
+.. code:: pycon
+
+    In [1]: a = np.arange(100, dtype=np.float64)
+
+    In [2]: %timeit myalgorithm(a, 1.0)
+    473 ns ± 11.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
+
+    In [3]: a = np.arange(100)
+
+    In [4]: %timeit myalgorithm(a, 1)
+    237 ns ± 4.59 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
+
+Writing Tests
+=============
+
+The following is a sufficient piece of test code for the overloaded
+``set_to_x`` implementation. As you can see, only a small part of the test-code
+is about testing if the function works correctly. Most of the test code in this
+example checks that all error cases are handled and that all raised exceptions
+are of the correct type and have the correct error message. When implementing
+tests cases for Numba, you should always use ``numba.tests.support.TestCase``.
+
+.. literalinclude:: test_myjitmodule2.py
+
+While it is a pretty decent test, it wouldn't be accepted into Numba.
+There are a few more test-cases that should be implemented:
+
+* Empty array
+* Single value array
+* Multidimensional arrays
+* Tests for ``int32`` and ``float32`` types
+
+You can download the example code from above including tests here:
+
+* :download:`mymodule.py </extending/mymodule.py>`
+* :download:`myjitmodule.py </extending/myjitmodule.py>`
+* :download:`myjitmodule2.py </extending/myjitmodule2.py>`
+* :download:`test_myjitmodule2.py </extending/test_myjitmodule2.py>`
+
+Implementing ``overload`` for NumPy functions
+=============================================
+
+Numba has strong support for NumPy because it provides ``jit`` compatible
+re-implementations of NumPy functions. In such cases ``overload`` is a very
+convenient option, however there are a few additional things to watch out for.
+
+* The Numba implementation should match the NumPy implementation as closely as
+  feasible with respect to accepted types, arguments, raised exceptions and
+  runtime (Big-O / Landau order).
+
+* When implementing supported argument types, bear in mind that---thanks to
+  duck typing---NumPy does tend to accept a multitude of argument types beyond
+  NumPy arrays such as scalar, list, tuple, set, iterator, generator etc.. So
+  you will need to account for that during type inference and subsequently as
+  part of the tests.
+
+* A NumPy function may return a scalar, array or a data structure
+  which matches one of its inputs - so you need to watch out for type
```

Comment:
link to http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-has-a-type-unification-problem ?

Reply:
Needs addressing.